itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcrossInstances.java

```diff
@@ -1321,6 +1322,106 @@ public Boolean apply(@Nullable CallerArguments args) {
+
+    tuple = primary.run("use " + primaryDbName)
+            .run("insert overwrite table t1 select * from t2")
+            .dump(primaryDbName, tuple.lastReplicationId);
+
+    testMoveOptimization(primaryDbName, replicatedDbName, replicatedDbName_CM, "t1", "INSERT", tuple);
+  }
+
+  @Test
+  public void testMoveOptimizationIncrementalFailureAfterCopy() throws Throwable {
+    List<String> withConfigs = Arrays.asList("'hive.repl.enable.move.optimization'='true'");
+    String replicatedDbName_CM = replicatedDbName + "_CM";
+    WarehouseInstance.Tuple tuple = primary.run("use " + primaryDbName)
+            .run("create table t2 (place string) partitioned by (country string)")
+            .run("ALTER TABLE t2 ADD PARTITION (country='india')")
+            .dump(primaryDbName, null);
+    replica.load(replicatedDbName, tuple.dumpLocation, withConfigs);
+    replica.load(replicatedDbName_CM, tuple.dumpLocation, withConfigs);
+    replica.run("alter database " + replicatedDbName + " set DBPROPERTIES ('" + SOURCE_OF_REPLICATION + "' = '1,2,3')")
+        .run("alter database " + replicatedDbName_CM + " set DBPROPERTIES ('" + SOURCE_OF_REPLICATION + "' = '1,2,3')");
+
+    tuple = primary.run("use " + primaryDbName)
+            .run("insert into table t2 partition(country='india') values ('bangalore')")
+            .dump(primaryDbName, tuple.lastReplicationId);
+
+    testMoveOptimization(primaryDbName, replicatedDbName, replicatedDbName_CM, "t2", "INSERT", tuple);
+  }
+
+  private void testMoveOptimization(String primarydb, String replicadb, String replicatedDbName_CM,
+                                    String tbl,  String eventType, WarehouseInstance.Tuple tuple) throws Throwable {
+    List<String> withConfigs = Arrays.asList("'hive.repl.enable.move.optimization'='true'");
+
+    // fail add notification for given event type.
+    BehaviourInjection<NotificationEvent, Boolean> callerVerifier
+            = new BehaviourInjection<NotificationEvent, Boolean>() {
+      @Nullable
+      @Override
+      public Boolean apply(@Nullable NotificationEvent entry) {
+        if (entry.getEventType().equalsIgnoreCase(eventType) && entry.getTableName().equalsIgnoreCase(tbl)) {
+          injectionPathCalled = true;
+          LOG.warn("Verifier - DB: " + String.valueOf(entry.getDbName())
+                  + " Table: " + String.valueOf(entry.getTableName())
+                  + " Event: " + String.valueOf(entry.getEventType()));
+          return false;
+        }
+        return true;
+      }
+    };
+
+    InjectableBehaviourObjectStore.setAddNotificationModifier(callerVerifier);
+    replica.loadFailure(replicadb, tuple.dumpLocation, withConfigs);
+    InjectableBehaviourObjectStore.resetAddNotificationModifier();
+    callerVerifier.assertInjectionsPerformed(true, false);
+    replica.load(replicadb, tuple.dumpLocation, withConfigs);
+
+    replica.run("use " + replicadb)
+            .run("select country from " + tbl)
```

Comment:
Can be select place from tbl where country='india' to we know if the data file is correct.

Reply:
done